File: code/ch04/4.2.1.24game.java
package ch04.code;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class TwentyFourGame {
    List<List<Double>> permutationList = new ArrayList<>();

    // 9ms at best
    public boolean judgePoint24(int[] nums) {
        permuteUnique(nums);
        for (List<Double> permutation : permutationList) {
            if (compute(permutation))
                return true;
        }
        return false;
    }

    private boolean compute(List<Double> nums) {
        if (nums.size() == 1)
            return Math.abs(nums.get(0) - 24) <= 1e-6;

        for (int i = 0; i < nums.size() - 1; i++) {
            // compute possible result from + - * /
            List<Double> tmpResultList = new ArrayList<>();
            tmpResultList.add(nums.get(i) + nums.get(i + 1));
            tmpResultList.add(nums.get(i) - nums.get(i + 1));
            tmpResultList.add(nums.get(i) * nums.get(i + 1));
            if (nums.get(i + 1) != 0) {
                tmpResultList.add(nums.get(i) / nums.get(i + 1));
            }

            // replace nums[i] and nums[i+1] with the result
            // continue with the new list
            for (Double newNum : tmpResultList) {
                List<Double> newList = new ArrayList<>(nums);
                newList.set(i, newNum);
                newList.remove(i + 1);
                if (compute(newList))
                    return true;
            }
        }
        return false;
    }


    public void permuteUnique(int[] nums) {
        boolean[] visited = new boolean[nums.length];
        Arrays.sort(nums);
        backtracking(nums, new ArrayList<>(), visited);
    }

    private void backtracking(int[] nums, List<Double> tmp, boolean[] visited) {
        if (tmp.size() == nums.length) {
            permutationList.add(new ArrayList<>(tmp));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (visited[i])
                continue;
            // !visited[i - 1] 说明已经遍历完被撤销了状态记录
            if (i > 0 && nums[i] == nums[i - 1] && !visited[i - 1])
                continue;
            visited[i] = true;
            tmp.add((double) nums[i]);
            backtracking(nums, tmp, visited);

            visited[i] = false;
            tmp.remove(tmp.size() - 1);
        }
    }
}

File: code/ch04/4.2.1.24game.py
from typing import List


class Solution:
    def judgePoint24(self, nums: List[int]) -> bool:
        permutations = self.permuteUnique(nums)
        for permutation in permutations:
            if self.compute(permutation):
                return True
        return False

    def compute(self, nums: List[float]) -> bool:
        if len(nums) == 1:
            return abs(nums[0] - 24) <= 0.00001
        for i in range(len(nums) - 1):
            # compute possible result from + - * /
            tmp = []
            tmp.append(nums[i] + nums[i + 1])
            tmp.append(nums[i] - nums[i + 1])
            tmp.append(nums[i] * nums[i + 1])
            if nums[i + 1] != 0:
                tmp.append(nums[i] / nums[i + 1])

            for num in tmp:
                new_list = nums[:]
                new_list[i] = num
                new_list.pop(i + 1)
                if self.compute(new_list):
                    return True
        return False

    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        permutations = []
        nums.sort()
        tmp = []
        visited = [False] * len(nums)

        self.backtracking(nums, tmp, visited, permutations)
        return permutations

    def backtracking(
        self, nums: List[int], tmp: List[float], visited: List[bool], perm: List[int],
    ) -> None:
        if len(nums) == len(tmp):
            perm.append(tmp[:])
            return
        for i in range(len(nums)):
            if visited[i]:
                continue
            if i > 0 and nums[i] == nums[i - 1] and not visited[i - 1]:
                continue
            visited[i] = True
            tmp.append(nums[i])
            self.backtracking(nums, tmp, visited, perm)
            visited[i] = False
            tmp.pop()
